rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py

```diff
@@ -0,0 +1,187 @@
+
+from ray.rllib.utils.tf_utils import (
+    explained_variance,
+    warn_if_infinite_kl_divergence,
+)
+
+from ray.rllib.utils.typing import TensorType
+
+tf1, tf, tfv = try_import_tf()
+
+logger = logging.getLogger(__name__)
+
+
+class PPOTfPolicyWithRLModule(
+    LearningRateSchedule,
+    EntropyCoeffSchedule,
+    KLCoeffMixin,
+    EagerTFPolicyV2,
+):
+    """PyTorch policy class used with PPO.
+
+    This class is copied from PPOTFPolicy and is modified to support RLModules.
+    Some subtle differences:
+    - if config._enable_rl_module api is true make_rl_module should be implemented by
+    the policy the policy is assumed to be compatible with rl_modules (i.e. self.model
+    would be an RLModule)
+    - Tower stats no longer belongs to the model (i.e. RLModule) instead it belongs to
+    the policy itself.
+    - Connectors should be enabled to use this policy
+    - So far it only works for vectorized obs and action spaces (Fully connected neural
+    networks). we need model catalog to work for other obs and action spaces.
+
+    # TODO: In the future we will deprecate doing all phases of training, exploration,
+    # and inference via one policy abstraction. Instead, we will use separate
+    # abstractions for each phase. For training (i.e. gradient updates, given the
+    # sample that have been collected) we will use RLTrainer which will own one or
+    # possibly many RLModules, and RLOptimizer. For exploration, we will use RLSampler
+    # which will own RLModule, and RLTrajectoryProcessor. The exploration and inference
+    # phase details are TBD but the whole point is to make rllib extremely modular.
+    """
+
+    def __init__(self, observation_space, action_space, config):
+        config = dict(ray.rllib.algorithms.ppo.ppo.PPOConfig().to_dict(), **config)
+        # TODO: Move into Policy API, if needed at all here. Why not move this into
+        #  `PPOConfig`?.
+        validate_config(config)
+        EagerTFPolicyV2.enable_eager_execution_if_necessary()
+        EagerTFPolicyV2.__init__(self, observation_space, action_space, config)
+        # Initialize MixIns.
+        LearningRateSchedule.__init__(self, config["lr"], config["lr_schedule"])
+        EntropyCoeffSchedule.__init__(
+            self, config["entropy_coeff"], config["entropy_coeff_schedule"]
+        )
+        KLCoeffMixin.__init__(self, config)
+
+        self.maybe_initialize_optimizer_and_loss()
+
+    @override(EagerTFPolicyV2)
+    def loss(
+        self,
+        model: Union[ModelV2, "tf.keras.Model"],
+        dist_class,
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        del dist_class
+        fwd_out = model.forward_train(train_batch)
+
+        curr_action_dist = fwd_out[SampleBatch.ACTION_DIST]
+        dist_class = curr_action_dist.__class__
+        value_fn_out = fwd_out[SampleBatch.VF_PREDS]
+
+        reduce_mean_valid = tf.reduce_mean
```

Comment:
what's the point of saving this in a local variable?

Reply:
I got rid of this because of the redundancy